### PR TITLE
Add --dryRun option

### PR DESCRIPTION
**Pull request by CyberTailor**
I'm not able to test this because nimble fails to build (`tools.nim(11, 14) Error: cannot open file: compiler/nimblecmd`) even with Nim compiled from devel branch, so testing is required.

This PR closes issue #261

### CODE REVIEW

File: tests/tuninstall.nim [hunk 1]
```diff
@@ -42,6 +47,45 @@ suite "uninstall":
...
+    check execNimbleYes("--dryRun", "uninstall", "packagea").exitCode == QuitSuccess
+    check execNimbleYes("--dryRun", "uninstall", "PackageA").exitCode == QuitSuccess
+
+    check execNimbleYes("--dryRun", "uninstall", "PackageB").exitCode == QuitSuccess
+
+    check execNimbleYes("--dryRun", "uninstall", "PackageA@0.2", "issue27b").exitCode !=
+        QuitSuccess
+
+    check execNimbleYes("--dryRun", "uninstall", "nimscript").exitCode == QuitSuccess
+
```
Comment (dom96) [inline tests/tuninstall.nim:88]: Does this test to verify that a package wasn't uninstalled due to the `--dryRun`? Maybe I'm missing it but I don't see it. Please call out the part of this test that tests for this with a comment in the code.

File: src/nimblepkg/nimscriptexecutor.nim [hunk 2]
```diff
@@ -38,7 +38,8 @@ proc execCustom*(nimbleFile: string, options: Options,
     raise nimbleError("Pre-hook prevented further execution.")
 
   if not nimbleFile.isNimScript(options):
-    writeHelp()
+    writeHelp(false)
+    raise nimbleQuit(QuitFailure)
```
Comment (dom96) [inline src/nimblepkg/nimscriptexecutor.nim:42]: Doesn't `writeHelp` already raise?

File: src/nimble.nim [hunk 3]
```diff
@@ -1738,7 +1765,8 @@ when fileExists("{nimblePathsFileName}"):
     writeFile = true
 
   if writeFile:
-    writeFile(nimbleConfigFileName, fileContent)
+    if not options.dryRun:
+      writeFile(nimbleConfigFileName, fileContent)
     displayInfo(&"\"{nimbleConfigFileName}\" is set up.")
```
Comment (dom96) [inline src/nimble.nim:1770]: seeing this with `--dryRun` would be quite confusing

File: src/nimble.nim [hunk 4]
```diff
@@ -1263,7 +1282,8 @@ proc updatePathsFile(pkgInfo: PackageInfo, options: Options) =
   for path in paths:
     pathsFileContent &= &"--path:{path.escape}\n"
   var action = if fileExists(nimblePathsFileName): "updated" else: "generated"
-  writeFile(nimblePathsFileName, pathsFileContent)
+  if not options.dryRun:
+    writeFile(nimblePathsFileName, pathsFileContent)
   displayInfo(&"\"{nimblePathsFileName}\" is {action}.")
```
Comment (dom96) [inline src/nimble.nim:1287]: this will seem confusing with --dryRun

**Review by dom96 — CHANGES_REQUESTED** [review]
This looks good, but I fear it will be quite fragile. It's unclear which procs handle `dryRun` and which don't. For example you have `setupNimbleConfig` handling it, but then below where it's called you see another check for `dryRun`. This means we might end up with `dryRun` not being correctly handled in the future because someone not familiar with the code forgets to use it, or uses it more than necessary.

It's definitely non-trivial to make this safer though unfortunately. The only thing I can think of is to create Nimble-specific wrappers for all procs that write/remove/etc files and execute commands. Tbh maybe we should do this? We already have `doCmd`.

My other concern is that we will see messages the imply actions were taken, even with the `dryRun`. We need to make it clear that they weren't taken.

**Comment by CyberTailor** [comment]
Closing as this implementation is too fragile

**Pull request closed without merging**